server/oauth_migration.go

```diff
@@ -0,0 +1,110 @@
+package main
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/mattermost/mattermost-plugin-api/cluster"
+	"github.com/pkg/errors"
+)
+
+const (
+	oauthMigrationStoreKey = "oauth_migration"
+	oauthMigrationMutexKey = "oauth_migration_mutex"
+
+	oauthMigrationStatusInProgress = "IN_PROGRESS"
+	oauthMigrationStatusComplete   = "COMPLETE"
+	oauthMigrationStatusError      = "ERROR"
+)
+
+func (p *Plugin) checkAndPerformOAuthTokenMigration() error {
+	mutexAPI := cluster.MutexPluginAPI(p.API)
+	mutex, err := cluster.NewMutex(mutexAPI, oauthMigrationMutexKey)
+	if err != nil {
+		return err
+	}
+
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	var status string
+	err = p.client.KV.Get(oauthMigrationStoreKey, &status)
+	if err != nil {
+		return err
+	}
+
+	// Migration is in progress or already completed
+	if status != "" {
+		return nil
+	}
+
+	_, _ = p.client.KV.Set(oauthMigrationStoreKey, oauthMigrationStatusInProgress)
+
+	err = p.notifyAllConnectedUsersToReconnect()
+	if err != nil {
+		_, _ = p.client.KV.Set(oauthMigrationStoreKey, oauthMigrationStatusError)
+		return err
+	}
+
+	_, _ = p.client.KV.Set(oauthMigrationStoreKey, oauthMigrationStatusComplete)
+
+	return nil
+}
+
+func (p *Plugin) notifyAllConnectedUsersToReconnect() error {
+	allKeys := []string{}
+	page := 0
+	for {
+		keys, err := p.client.KV.ListKeys(page, 100)
+		if err != nil {
+			return errors.Wrap(err, "error listing keys for connected users")
+		}
+
+		if len(keys) == 0 {
+			break
+		}
+
+		keysToAdd := []string{}
+		for _, key := range keys {
+			if strings.HasSuffix(key, GitlabTokenKey) {
+				keysToAdd = append(keysToAdd, key)
+			}
+
+			if strings.HasSuffix(key, GitlabUsernameKey) || strings.HasSuffix(key, GitlabIDUsernameKey) {
+				_ = p.client.KV.Delete(key)
```

Comment:
I actually don't think we should delete these keys. Incoming webhooks can still successfully show the connected mentions this way, before the user reconnects their account.